File: CombinationSum/I/CombinationSum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> result;
        vector<int> tmp;
        getresult(result, tmp, candidates, target, 0);
        return result; 
    }
private:
	void getresult(vector<vector<int>> &result, vector<int> &tmp, const vector<int> candidates, int target, int begin) {
		if(target == 0) {
			result.push_back(tmp);
			return;
		}
		for(int i = begin; i < candidates.size() && target >= candidates[i]; i++) {
			tmp.push_back(candidates[i]);
			getresult(result, tmp, candidates, target - candidates[i], i);
			tmp.pop_back();
		}
	}
};

File: CombinationSum/I/CombinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        res = []
        sol = []
        getSubsol(candidates, target, res, sol)
        return res

def getSubsol(candidates, target, res, sol):
    for i in range(len(candidates)):
        if target == candidates[i]:
            res.append(sol + [candidates[i]])
            return
        if target < candidates[i]:
            break
        else:
            getSubsol(candidates[i:], target - candidates[i],res, sol + [candidates[i]])